### ISSUE TITLE
Introduce async API

### ISSUE THREAD

**Issue by AlekseyMartynov** [DevExpress/DevExtreme.AspNet.Data#337]
Started in #308

Try to detect and use async Queryable extensions of LINQ providers.

**Comment by AlekseyMartynov** [comment]
**EF 6**
```c#
// System.Data.Entity.QueryableExtensions
var count = await dbSet.CountAsync();
var array = await dbSet.ToArrayAsync();

// System.Data.Entity.Infrastructure
var enumerable = (IDbAsyncEnumerable<DataItem>)dbSet;
var enumerator = enumerable.GetAsyncEnumerator();
while(await enumerator.MoveNextAsync(CancellationToken.None)) {
    var item = enumerator.Current;
}
```

**EF Core**
```c#
// Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions
var count = await dbSet.CountAsync();
var array = await dbSet.ToArrayAsync();

// Microsoft.EntityFrameworkCore.Query.Internal
var accessor = (IAsyncEnumerableAccessor<DataItem>)dbSet;
var enumerator = accessor.AsyncEnumerable.GetEnumerator();
while(await enumerator.MoveNext(CancellationToken.None)) {
    var item = enumerator.Current;
}
```

**NHibernate**
```c#
// NHibernate.Linq.LinqExtensionMethods
var count = await query.CountAsync();
var list = await query.ToListAsync();

var future = query.ToFuture();
var ordinaryEnumerable = await future.GetEnumerableAsync(); // is List<DataItem>
```

**XPO**
```c#
// DevExpress.Xpo.XPQueryExtensions
var count = await query.CountAsync();
var array = await query.ToArrayAsync();

var ordinaryEnumerable = await query.EnumerateAsync(); // is Cast<T> over ICollection

// weird non-awaitable version
query.EnumerateAsync((IEnumerable<DataItem> result, Exception x) => {

});
```

NOTES
- EF and EF Core provide per-item async enumeration, however namespace names suggest that this API is internal
- NH and XPO provide async methods returning `IEnumerable`, however the actual object is `ICollection`.

**Comment by Bykiev** [comment]
@AlekseyMartynov, is any estimate date when it'll be released?

**Comment by AlekseyMartynov** [comment]
@Bykiev 
You asked at the right moment.
https://github.com/DevExpress/DevExtreme.AspNet.Data/releases/tag/2.5.0-rc1

**Comment by Bykiev** [comment]
@AlekseyMartynov, just checked and it's working fine with EF 6.2.0. Thanks a lot!

**Comment by Bykiev** [comment]
@AlekseyMartynov, do you have any estimate release date of this feature?

**Comment by AlekseyMartynov** [comment]
@Bykiev During September.

**Comment by AlekseyMartynov** [comment]
Released in [2.5.0](https://github.com/DevExpress/DevExtreme.AspNet.Data/releases/tag/2.5.0).
